lib/scholar/neighbors/kd_tree.ex

```diff
@@ -0,0 +1,318 @@
+defmodule Scholar.Neighbors.KDTree do
+  @moduledoc """
+  Implements a kd-tree, a space-partitioning data structure for organizing points
+  in a k-dimensional space.
+
+  This is implemented as one-dimensional tensor with indices pointed to highest
+  dimension of the given tensor. Traversal starts by calling `root/0` and then
+  accessing the `left_child/1` and `right_child/1`. The tree is left-balanced.
+
+  Two construction modes are available:
+
+    * `banded/2` - the tensor has min and max values with an amplitude given by `max - min`.
+      It is also guaranteed that the `amplitude * levels(tensor) + 1` does not overflow
+      the tensor. See `amplitude/1` to verify if this holds. This implementation happens
+      fully within `defn`.
+
+    * `unbanded/2` - there are no known bands (min and max values) to the tensor.
+      This implementation is recursive and goes in and out of the `defn`, therefore
+      it cannot be called inside `defn`.
+
+  ## References
+
+    * [GPU-friendly, Parallel, and (Almost-)In-Place Construction of Left-Balanced k-d Trees](https://arxiv.org/pdf/2211.00120.pdf).
+  """
+
+  import Nx.Defn
+
+  # TODO: Benchmark
+
+  @derive {Nx.Container, keep: [:levels], containers: [:indexes]}
+  @enforce_keys [:levels, :indexes]
+  defstruct [:levels, :indexes]
+
+  @doc """
+  Builds a KDTree without known min-max bounds.
+
+  If your tensor has a known bound (for example, -1 and 1),
+  consider using the `banded/2` version which is more efficient.
+
+  ## Options
+
+    * `:compiler` - the default compiler to use for internal defn operations
+
+  ## Examples
+
+      iex> Scholar.Neighbors.KDTree.unbanded(Nx.iota({5, 2}), compiler: EXLA.Defn)
+      %Scholar.Neighbors.KDTree{
+        levels: 3,
+        indexes: Nx.u32([3, 1, 4, 0, 2])
+      }
+
+  """
+  def unbanded(tensor, opts \\ []) do
+    levels = levels(tensor)
+    {size, _dims} = Nx.shape(tensor)
+
+    indexes =
+      if size > 2 do
+        subtree_size = unbanded_subtree_size(1, levels, size)
+        {left, mid, right} = Nx.Defn.jit_apply(&root_slice(&1, subtree_size), [tensor], opts)
+
+        acc = <<Nx.to_number(mid)::32-unsigned-native-integer>>
+        acc = recur([{1, left}, {2, right}], [], acc, tensor, 1, levels, opts)
+        Nx.from_binary(acc, :u32)
+      else
+        Nx.argsort(tensor[[.., 0]], direction: :desc, type: :u32)
+      end
+
+    %__MODULE__{levels: levels, indexes: indexes}
+  end
+
+  defp recur([{_i, %Nx.Tensor{shape: {1}} = leaf} | rest], next, acc, tensor, level, levels, opts) do
+    [leaf] = Nx.to_flat_list(leaf)
```

Comment:
Any special reason to use `Nx.to_flat_list` instead of `Nx.to_number` here?

Reply:
to_number only works on scalars.